### ISSUE TITLE
Blueprint#initialize; full stack trace array not assigned to ivar @caller

### ISSUE THREAD

**Issue by carrola7** [jbodah/spy_rb#52]
Hey Josh,

First thing I wanted to run by you was the `initialize` method in `blueprint.rb`.

```ruby
module Spy
  class Blueprint
    attr_reader :target, :msg, :type

    def initialize(target, msg, type)
      @target = target
      @msg = msg
      @type = type
      @caller = _caller
    end

    alias :_caller :caller

    def caller
      @caller
    end

    def to_s
      [@target.object_id, @msg, @type].join("|")
    end
  end
end
```

I was sticking a few `binding.pry`s around the place and noticed that after aliasing `_caller`, it's value was an array containing the stack trace, as expected. On `line 9` this value is assigned to `@caller` but if you check the value of `@caller` after this it only contains the first two elements from the `_caller` array. Why is the full array not assigned to `@caller`? 

Maybe this isn't important but it would be nice to shed some light on this and also on a high level what is the benefit of doing this?

Cheers

**Comment by jbodah** [comment]
Adding a `puts @caller` after the assignment in the initialize call dumps the full stacktrace for me (just running the tests below:

```
.["/Users/jbodah/repos/spy_rb/lib/spy/api.rb:25:in `new'", "/Users/jbodah/repos/spy_rb/lib/spy/api.rb:25:in `on'", "/Users/jbodah/repos/spy_rb/test/spy_test.rb:126:in `block (4 levels) in <class:SpyTest>'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:108:in `block (3 levels) in run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:206:in `capture_exceptions'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:105:in `block (2 levels) in run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:258:in `time_it'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:104:in `block in run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:320:in `on_signal'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:278:in `with_info_handler'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest/test.rb:103:in `run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:767:in `run_one_method'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:294:in `run_one_method'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:288:in `block (2 levels) in run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:287:in `each'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:287:in `block in run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:320:in `on_signal'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:307:in `with_info_handler'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:286:in `run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:150:in `block in __run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:150:in `map'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:150:in `__run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:127:in `run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-tagz-1.6.0/lib/minitest/tagz.rb:56:in `run'", "/usr/local/lib/ruby/gems/2.5.0/gems/minitest-5.4.3/lib/minitest.rb:56:in `block in autorun'"]
.

Fabulous run in 0.130529s, 574.5850 runs/s, 1126.1865 assertions/s.
```

The only thing caching `caller` does for us is it lets us tell the user where we think the previous `Spy.on` call is on errors: https://github.com/jbodah/spy_rb/commit/1b9f9cb33c013e20d1608d2671e01708b12b1d30#diff-e0b1369de2415ed1ca7ce1724a4091d4R24. It's minor but comes in handy sometimes

`Blueprint` was just added to encapsulate the thing that the user asked spy to modify. The main thing it abstracts away is the serialization strategy (defined by `#to_s`). This cleaned up [Spy::Registry](https://github.com/jbodah/spy_rb/blob/master/lib/spy/registry.rb#L16) a lot so it's just a light hash wrapper. [Here's an older version of that same code](https://github.com/jbodah/spy_rb/blob/dddef5b9838e1351c79c43c7b0c58b7bb3f2bba4/lib/spy/collection.rb) which got hard to follow and extend

**Comment by jbodah** [comment]
One thought why you might be hitting that is that `caller` acts different when called in a pry session(? `caller` is with respect to your pry repl rather than from the initialize method)
